gateway/ln-gateway/src/events.rs

```diff
@@ -22,4 +32,288 @@ pub const ALL_GATEWAY_EVENTS: [EventKind; 11] = [
+    client: &Arc<ClientHandle>,
+    start: SystemTime,
+    end: SystemTime,
+) -> Vec<PersistedLogEntry> {
+    const BATCH_SIZE: u64 = 10_000;
+    let mut end_position = {
+        let mut dbtx = client.db().begin_transaction_nc().await;
+        dbtx.get_next_event_log_id().await
+    };
+
+    let mut start_position = end_position.saturating_sub(BATCH_SIZE);
+    let mut all_events = Vec::new();
+
+    let log_start = EventLogId::new(0);
+
+    // Read the event log until we hit the end.
+    while start_position != log_start {
+        let batch = client.get_event_log(Some(start_position), BATCH_SIZE).await;
+        let (start_index, end_index) = get_index_bounds(&batch, start, end);
+
+        // If the `end_index` is non-zero, that means that all events < `end_index` are
+        // within our time bounds
+        if end_index > 0 {
+            let events = &batch[start_index..end_index];
+            all_events.append(&mut events.to_vec());
+
+            // If the `start_index` is non-zero, we have found the last event to be
+            // included. Return all events.
+            if start_index > 0 {
+                return all_events;
+            }
+        }
+
+        end_position = start_position;
+        start_position = start_position.saturating_sub(BATCH_SIZE);
+    }
+
+    let remaining_log = end_position.distance(start_position);
+    if let Some(batch_size) = remaining_log {
+        let batch = client.get_event_log(Some(start_position), batch_size).await;
+        let (start_index, end_index) = get_index_bounds(&batch, start, end);
+
+        if end_index > 0 {
+            let events = &batch[start_index..end_index];
+            all_events.append(&mut events.to_vec());
+        }
+    }
+
+    all_events
+}
+
+/// Binary searches the `PersistedLogEntry` slice for the start timestamp and
+/// end timestamp and returns the index of each. If an exact match is not found,
+/// the index where start or end could be inserted to maintain ordering
+/// is returned.
+fn get_index_bounds(
+    batch: &[PersistedLogEntry],
+    start: SystemTime,
+    end: SystemTime,
+) -> (usize, usize) {
+    let timestamps = batch.iter().map(|e| e.timestamp).collect::<Vec<_>>();
+    let start_micros = start
+        .duration_since(UNIX_EPOCH)
+        .expect("before unix epoch")
+        .as_micros() as u64;
+    let index = timestamps.binary_search(&start_micros);
+    let start_index = match index {
+        Err(index) | Ok(index) => index,
+    };
+
+    let end_micros = end
+        .duration_since(UNIX_EPOCH)
+        .expect("before unix epoch")
+        .as_micros() as u64;
+    let index = timestamps.binary_search(&end_micros);
+    let end_index = match index {
+        Err(index) | Ok(index) => index,
+    };
+
+    (start_index, end_index)
+}
+
+/// Filters the given `PersistedLogEntry` slice by the `EventKind` and
+/// `ModuleKind`.
+pub(crate) fn filter_events(
+    all_events: &[PersistedLogEntry],
```

Comment:
If you take it as a `IntoIter<Item = PersistedLogEntry>` and return `-> impl Iterator<Item = PersistedLogEntry` you can make the whole thing never even allocate extra.

Reply:
I am having difficulty with this. If I take `all_events` as `IntoIter<Item = PersistedLogEntry>`, won't that consume `all_events`? I want to call this multiple times over the same slice. I think copying it between calls would be expensive, since it contains all of the events from the time duration.

Reply:
@m1sterc001guy https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fc0641fe97ed45728a95a163bd2b9faf